File: appledoc/Parsing/Parser.m
//
//  Parser.m
//  appledoc
//
//  Created by Tomaž Kragelj on 3/19/12.
//  Copyright (c) 2012 Tomaz Kragelj. All rights reserved.
//

#import "Objects.h"
#import "Store.h"
#import "ObjectiveCParser.h"
#import "Parser.h"

typedef void(^ParserPathBlock)(NSString *path);

#pragma mark - 

@implementation Parser

#pragma mark - Task invocation

- (NSInteger)runTask {
	LogNormal(@"Starting parsing...");
	__weak Parser *blockSelf = self;
	__block GBResult result = GBResultOk;
	[self.settings.arguments enumerateObjectsUsingBlock:^(NSString *path, NSUInteger idx, BOOL *stop) {
		GBResult pr = [blockSelf parsePath:path withBlock:^(NSString *path) {
			if (![blockSelf isSourceCodeFile:path]) return;
			LogInfo(@"Parsing source file '%@'...", path);
			NSInteger parseResult = [self.objectiveCParser parseFile:path withSettings:self.settings store:self.store];
			if (parseResult > result) result = parseResult;
		}];
		if (pr > result) result = pr;
	}];
	LogInfo(@"Parsing finished.");
	return result;
}

#pragma mark - Parsing helpers

- (NSInteger)parsePath:(NSString *)path withBlock:(ParserPathBlock)handler {
	LogVerbose(@"Parsing '%@'...", path);
	NSString *standardized = [path gb_stringByStandardizingCurrentDirAndPath];
	
	if (![self.fileManager fileExistsAtPath:standardized]) {
		LogWarn(@"'%@' doesn't exist, skipping!", path);
		return GBResultSystemError;
	}
	
	if ([self.fileManager gb_fileExistsAndIsDirectoryAtPath:standardized]) {
		[self parseDirectory:standardized withBlock:handler];
		return GBResultOk;
	}
	
	[self parseFile:standardized withBlock:handler];
	return GBResultOk;
}

- (NSInteger)parseDirectory:(NSString *)path withBlock:(ParserPathBlock)handler {
	if ([self isPathIgnored:path]) {
		LogNormal(@"Path '%@' ignored, skipping...", path);
		return GBResultOk;
	}	
	LogVerbose(@"Parsing directory '%@'...", path);

	// Get contents of the directory.
	NSError *error = nil;
	NSArray *contents = [self.fileManager contentsOfDirectoryAtPath:path error:&error];
	if (error) {
		LogNSError(error, @"Failed fetching contents of '%@'!", path);
		return GBResultSystemError;
	}
	
	// First files...
	[contents enumerateObjectsUsingBlock:^(NSString *subpath, NSUInteger idx, BOOL *stop) {
		NSString *fullPath = [path stringByAppendingPathComponent:subpath];
		if (![self.fileManager gb_fileExistsAndIsFileAtPath:fullPath]) return;
		if ([self isFileIgnored:subpath]) return;
		[self parseFile:fullPath withBlock:handler];
	}];
	
	// ...then directories.
	[contents enumerateObjectsUsingBlock:^(NSString *subpath, NSUInteger idx, BOOL *stop) {
		NSString *fullPath = [path stringByAppendingPathComponent:subpath];
		if (![self.fileManager gb_fileExistsAndIsDirectoryAtPath:fullPath]) return;
		if ([self isDirectoryIgnored:subpath]) return;
		[self parseDirectory:fullPath withBlock:handler];
	}];
	
	return GBResultOk;
}

- (NSInteger)parseFile:(NSString *)path withBlock:(ParserPathBlock)handler {
	if ([self isPathIgnored:path]) {
		LogNormal(@"Path '%@' ignored, skipping...", path);
		return GBResultOk;
	}
	LogVerbose(@"Parsing file '%@'...", path);
	handler(path);
	return GBResultOk;
}

#pragma mark - Helper methods

- (BOOL)isPathIgnored:(NSString *)path {
	__block BOOL result = NO;
	[self.settings.ignoredPaths enumerateObjectsUsingBlock:^(NSString *ignored, NSUInteger idx, BOOL *stop) {
		if ([path hasSuffix:ignored]) {
			result = YES;
			*stop = YES;
		}
	}];
	return result;
}

- (BOOL)isFileIgnored:(NSString *)filename {
	if ([filename isEqualToString:@".DS_Store"]) return YES;
	return NO;
}

- (BOOL)isDirectoryIgnored:(NSString *)filename {
	if ([filename isEqualToString:@".git"]) return YES;
	if ([filename isEqualToString:@".svn"]) return YES;
	if ([filename isEqualToString:@".hg"]) return YES;
	if ([filename hasSuffix:@".xcodeproj"]) return YES;
	return NO;
}

- (BOOL)isSourceCodeFile:(NSString *)path {
	NSString *extension = [path pathExtension];
	if ([extension isEqualToString:@"h"]) return YES;
	if ([extension isEqualToString:@"hh"]) return YES;
	if ([extension isEqualToString:@"m"]) return YES;
	if ([extension isEqualToString:@"mm"]) return YES;
	return NO;
}

#pragma mark - Properties

- (ParserTask *)objectiveCParser {
	if (_objectiveCParser) return _objectiveCParser;
	LogIntDebug(@"Initializing objective c parser due to first access...");
	_objectiveCParser = [[ObjectiveCParser alloc] init];
	return _objectiveCParser;
}

@end

File: appledoc/Parser/Parser.swift
//
//  Created by Tomaz Kragelj on 11.11.15.
//  Copyright © 2015 Gentle Bytes. All rights reserved.
//

import Foundation

/** Parses all source files in the command line arguments.
*/
class Parser: Task {
	
	// MARK: - Task
	
	/** Launches parsing tasks.
	
	Note that it's required to assign `settings` and `store` prior to calling this function!
	*/
	func run() throws {
		gtaskstart()
		ginfo("Parsing")
		
		for path in settings.arguments {
			let pathInfo = PathInfo(path: path as! String)
			try parsePathOrFile(pathInfo)
		}
		
		gtaskend()
	}
	
	// MARK: - Parsing handling
	
	fileprivate func parsePathOrFile(_ path: PathInfo) throws {
		if FileManager.default.isPathDirectory(path.fullPath) {
			gverbose("Scanning \(path.path)")
			try parseFolder(path)
		} else {
			gdebug("Scanning \(path.path)")
			try parseFile(path)
		}
	}
	
	fileprivate func parseFolder(_ path: PathInfo) throws {
		let subpathNames = try FileManager.default.contentsOfDirectory(atPath: path.fullPath)
		for subpathName in subpathNames {
			switch subpathName {
			case ".": break
			case "..": break
			default:
				let subpath = path.path.stringByAppendingPathComponent(subpathName)
				let subpathInfo = PathInfo(path: subpath)
				try parsePathOrFile(subpathInfo)
				break
			}
		}
	}
	
	fileprivate func parseFile(_ path: PathInfo) throws {
		switch path.path.pathExtension {
		case "h": try parseObjectiveCFromFile(path)
		case "m": try parseObjectiveCFromFile(path)
		case "mm": try parseObjectiveCFromFile(path)
		default: break // Ignore everything else
		}
	}
	
	fileprivate func parseObjectiveCFromFile(_ path: PathInfo) throws {
		gverbose("Parsing \(path) as Objective-C")
		objectiveCParser.path = path
		try objectiveCParser.run()
	}
	
	// MARK: - Derived properties
	
	lazy var objectiveCParser: ObjectiveCParser = {
		let result = ObjectiveCParser()
		result.settings = self.settings
		result.store = self.store
		return result
	}()
	
	// MARK: - Properties
	
	/// Application settings. This must be assigned prior to using the object!
	var settings: Settings!
	
	/// Application objects store. This must be assigned prior to using the object!
	var store: Store!
	
}
